### PR TITLE
Master owl aab

### PR DESCRIPTION
**Pull request by aab-odoo**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web/static/src/js/env.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+// for more information on environments.
+return {
+    _lt,
+    _t,
+    bus,
+    config,
+    qweb,
+    rpc: performRPC,
+    services: {
+        ajaxJsonRPC,
```
Comment (aab-odoo) [inline addons/web/static/src/js/env.js]: rpc methods should use (un)blockui fn from env (so they need to be more elaborate, see misc.js)

File: addons/web/static/src/js/main.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+const webClient = new WebClient();
+
+async function startWebClient() {
+    QWeb.DIRECTIVE_NAMES.inherit = 1;
+    QWeb.DIRECTIVE_NAMES['inherit-mode'] = 1;
+
+    await session.is_bound;
+    env.qweb.addTemplates(session.owlTemplates);
+    Component.env = env;
```
Comment (aab-odoo) [inline addons/web/static/src/js/main.js]: use aab version of main.js

File: addons/web/static/src/js/env.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+// There should be as much dependencies as possible in the env object.
+// This will allow an easier testing of components.
+// See https://github.com/odoo/owl/blob/master/doc/reference/environment.md#content-of-an-environment
+// for more information on environments.
+return {
+    _lt,
+    _t,
+    bus,
+    config,
```
Comment (aab-odoo) [inline addons/web/static/src/js/env.js]: replace config by its content (device, isDebug)

File: addons/web/static/src/js/main.js (hunk 2 above)
Comment (aab-odoo) [reply addons/web/static/src/js/main.js]: do `owl.config.mode = debug ? "dev" : "prod";` here instead of debugmanager

File: addons/web/static/tests/component_extension_tests.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+        class Parent extends Component {}
+        Parent.env = makeTestEnvironment({
+            rpc: () => Promise.resolve(),
+        });
+        Parent.template = xml`<div/>`;
+
+        const parent = new Parent();
+
+        parent.rpc({}).then(() => { throw new Error(); });
```
Comment (aab-odoo) [inline addons/web/static/tests/component_extension_tests.js]: const prom = ...
destroy()
await prom

File: addons/web/static/tests/component_extension_tests.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+const testUtils = require("web.test_utils");
+
+const { Component, tags } = owl;
+const { xml } = tags;
+
+QUnit.module("web", {
+    afterEach: function () {
+        // TODO: better than override destroy? (this was done for old widget)
+        testUtils.cleanTarget();
+    }
```
Comment (aab-odoo) [inline addons/web/static/tests/component_extension_tests.js]: remove this

File: addons/web/views/webclient_templates.xml [hunk 6]
```diff
@@ -262,6 +264,7 @@
         <script type="text/javascript" src="/web/static/src/js/tools/test_menus_loader.js"/>
         <script type="text/javascript" src="/web/static/src/js/tools/debug_manager_backend.js"></script>
         <script type="text/javascript" src="/web/static/src/js/tools/tools.js"></script>
+        <script type="text/javascript" src="/web/static/src/js/env.js"></script>
```
Comment (aab-odoo) [inline addons/web/views/webclient_templates.xml]: put an env.js file in assets frontend with services.rpc, qweb... (mandatory stuff)

File: addons/web/static/tests/helpers/test_env.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
+odoo.define("web.test_env", async function (require) {
```
Comment (aab-odoo) [inline addons/web/static/tests/helpers/test_env.js]: squash with prev

File: addons/web/static/tests/helpers/test_env.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+    // QWeb instance
+    const testQweb = new owl.QWeb({ templates: session.owlTemplates });
+
+    // TODO: we should not always add bus
+    // --> general problem: how to extend correctly the env
+    // use option like useBus, useConfig,... ?
+    // TODO: allow to use mockReadGroup,... ?
+    return {
+        bus: testBus,
```
Comment (aab-odoo) [inline addons/web/static/tests/helpers/test_env.js]: bus: new Bus(),

File: addons/web/static/tests/helpers/test_env.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+ */
+function makeTestEnvironment(env = {}) {
+    /**
+     * Helper allowing to prevent the use of some functions if they are not implemented
+     * first in the `env` object.
+     *
+     * @param {string} fnPath dot-separated path to designated property
+     * @param {Function} [callback] custom callback to be called as a wrapper: it
+     *      will take the implemented function as the first argument and will
+     *      transmit the remaining arguments.
```
Comment (aab-odoo) [inline addons/web/static/tests/helpers/test_env.js]: update docstring

File: addons/web/static/tests/helpers/test_env.js [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    // QWeb instance
+    const testQweb = new owl.QWeb({ templates: session.owlTemplates });
+
+    // TODO: we should not always add bus
+    // --> general problem: how to extend correctly the env
+    // use option like useBus, useConfig,... ?
+    // TODO: allow to use mockReadGroup,... ?
+    return {
+        bus: testBus,
+        config: testConfig,
```
Comment (aab-odoo) [inline addons/web/static/tests/helpers/test_env.js]: adapt to previous commit change (device, isDebug) + generate those objects here from scratch

File: addons/web/static/tests/helpers/test_env.js [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    }, env.services);
+
+    // Session
+    const testSession = Object.assign({}, session, {
+        // mandatory session functions
+        // ...
+    }, env.session);
+
+    // Translation
+    const testTranslate = '_t' in env ? env._t : _mandatory('_t');
```
Comment (aab-odoo) [inline addons/web/static/tests/helpers/test_env.js]: fallback on identity

Comment (aab-odoo) [reply addons/web/static/tests/helpers/test_env.js]: env._t || identity

File: addons/web/static/tests/helpers/test_env.js [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+        setCookie: _mandatory('services.setCookie'),
+        unblockUI: _mandatory('services.unblockUI'),
+        // ...
+    }, env.services);
+
+    // Session
+    const testSession = Object.assign({}, session, {
+        // mandatory session functions
+        // ...
+    }, env.session);
```
Comment (aab-odoo) [inline addons/web/static/tests/helpers/test_env.js]: from scratch

File: addons/web/static/tests/helpers/test_env.js [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+    const testBus = Object.create(bus, {
+        // mandatory bus functions
+        // ...
+    });
+
+    // Config
+    const testConfig = Object.assign({}, config, {
+        isDebug: _mandatory('config.isDebug'),
+        // ...
+    }, env.config);
```
Comment (aab-odoo) [inline addons/web/static/tests/helpers/test_env.js]: remove

File: addons/web/static/tests/helpers/test_env.js [hunk 14]
```diff
@@ -0,0 +1,93 @@
+odoo.define("web.test_env", async function (require) {
+"use strict";
+
+const { bus } = require("web.core");
+const config = require("web.config");
+const session = require("web.session");
+const rpc = require("web.rpc");
```
Comment (aab-odoo) [inline addons/web/static/tests/helpers/test_env.js]: remove bus config session, add Bus

File: addons/web_tour/static/src/js/running_tour_action_helper.js [hunk 15]
```diff
@@ -128,17 +132,31 @@ var RunningTourActionHelper = core.Class.extend({
...
+                ) {
+                    insertedText = String.fromCharCode(code);
+                }
             }
-            values.$element.trigger({type: "keyup", keyCode: keyCode});
+            values.$element.trigger(Object.assign({ type: "keydown" }, eventOptions));
+            if (insertedText) {
+                document.execCommand("insertText", 0, insertedText);
+            }
+            values.$element.trigger(Object.assign({ type: "keyup" }, eventOptions));
```
Comment (aab-odoo) [inline addons/web_tour/static/src/js/running_tour_action_helper.js]: do this in a separated commit

File: addons/web/static/tests/helpers/test_utils_dom.js [hunk 16]
```diff
@@ -12,6 +12,24 @@ var concurrency = require('web.concurrency');
...
+ */
+function returnAfterNextAnimationFrame() {
+    return new Promise((resolve, reject) => {
+        window.requestAnimationFrame(async () => {
+            await concurrency.delay(0);
+            resolve();
+        });
+    });
+
+    // return concurrency.delay(0);
```
Comment (aab-odoo) [inline addons/web/static/tests/helpers/test_utils_dom.js]: clean file

File: addons/web/static/tests/helpers/test_utils_dom.js [hunk 17]
```diff
@@ -12,6 +12,24 @@ var concurrency = require('web.concurrency');
...
+/**
+ * Returns a promise that will be resolved after the tick after the
+ * nextAnimationFrame
+ *
+ * This is useful to guarantee that OWL has had the time to render
+ *
+ * @returns {Promise}
+ */
+function returnAfterNextAnimationFrame() {
+    return new Promise((resolve, reject) => {
```
Comment (aab-odoo) [inline addons/web/static/tests/helpers/test_utils_dom.js]: unused arg, lint file

File: addons/web/static/tests/helpers/test_utils_dom.js [hunk 18]
```diff
@@ -81,52 +99,51 @@ function dragAndDrop($el, $to, options) {
...
         }
     } else {
         // It's impossible to drag another element when one is already
         // being dragged. So it's necessary to finish the drop when the test is
         // over otherwise it's impossible for the next tests to drag and
         // drop elements.
-        $el.on('remove', function () {
-            triggerEvent($el, 'mouseup');
+        $el.on('remove', async () => {
+            await triggerEvent($el, 'mouseup', {}, true);
```
Comment (aab-odoo) [inline addons/web/static/tests/helpers/test_utils_dom.js]: unnecessary diff

File: addons/web/static/tests/helpers/test_utils_dom.js [hunk 19]
```diff
@@ -197,64 +215,76 @@ function triggerKeypressEvent(char) {
...
+        throw new Error(`Element to click on is not visible ${selectorMsg}`);
     }
 
-    return concurrency.delay(0);
+    let event = new MouseEvent('click', {
+        view: window,
+        bubbles: true,
+        cancelable: true
+    });
+    target.dispatchEvent(event);
```
Comment (aab-odoo) [inline addons/web/static/tests/helpers/test_utils_dom.js]: use triggerEvent('click')

File: addons/web/static/tests/helpers/test_utils_dom.js [hunk 20]
```diff
@@ -266,64 +296,165 @@ function clickFirst(el, options) {
...
+    } else if (eventType === 'mouseleave') {
+        eventType = 'mouseout';
+    }*/
+
+    /*if (['mouseenter', 'mouseleave', 'mousemove', 'mouseover', 'mouseout', 'mouseup', 'mousedown'].includes(eventType)) {
+        const event = Object.keys(eventAttrs).length ?
+            $.Event(eventType, eventAttrs) :
+            eventType;
+        $(target).trigger(event);
+    } else {*/
```
Comment (aab-odoo) [inline addons/web/static/tests/helpers/test_utils_dom.js]: clean this

File: addons/web/static/tests/helpers/test_utils_dom.js [hunk 21]
```diff
@@ -266,64 +296,165 @@ function clickFirst(el, options) {
...
+        'keyup':{ constructor: KeyboardEvent , processParameters: (args) => { return Object.assign({}, args, { bubbles: true, keyCode: args.which }); }},
+
+        'drag':{ constructor: DragEvent , processParameters: onlyBubble },
+        'dragend':{ constructor: DragEvent , processParameters: onlyBubble },
+        'dragenter':{ constructor: DragEvent , processParameters: onlyBubble },
+        'dragstart':{ constructor: DragEvent , processParameters: onlyBubble },
+        'dragleave':{ constructor: DragEvent , processParameters: onlyBubble },
+        'dragover':{ constructor: DragEvent , processParameters: onlyBubble },
+        'drop':{ constructor: DragEvent , processParameters: onlyBubble },
+    };
```
Comment (aab-odoo) [inline addons/web/static/tests/helpers/test_utils_dom.js]: define all this outside (+no quotes in dict)

File: addons/web/static/tests/helpers/test_utils_dom.js [hunk 22]
```diff
@@ -266,64 +296,165 @@ function clickFirst(el, options) {
...
+
+    /*if (['mouseenter', 'mouseleave', 'mousemove', 'mouseover', 'mouseout', 'mouseup', 'mousedown'].includes(eventType)) {
+        const event = Object.keys(eventAttrs).length ?
+            $.Event(eventType, eventAttrs) :
+            eventType;
+        $(target).trigger(event);
+    } else {*/
+        if (!eventTypes[eventType] && !eventTypes[eventType.type])
+            event = new Event(eventType, Object.assign({}, eventAttrs, {bubbles: true}));
+        else {
```
Comment (aab-odoo) [inline addons/web/static/tests/helpers/test_utils_dom.js]: if () { }

File: addons/web/static/tests/helpers/test_utils_dom.js (hunk 20 above)
Comment (aab-odoo) [reply addons/web/static/tests/helpers/test_utils_dom.js]: lint file

File: addons/web/static/tests/helpers/test_utils_create.js [hunk 23]
```diff
@@ -395,14 +392,64 @@ function createParent(params) {
...
+/**
+ * Clears the current target of its added classes and event listeners. Note that
+ * you can call this function event if `prepareTarget` has not been called previously
+ */
+function cleanTarget() {
+    const { classList, events, el } = currentTarget;
+    if (el) {
+        classList.forEach(className => el.classList.remove(className));
+        for (const eventName in events) {
+            el.removeEventListener(eventName, events[eventName]);
```
Comment (aab-odoo) [inline addons/web/static/tests/helpers/test_utils_create.js]: check this (please no handler on body)

File: addons/web/static/tests/helpers/test_utils_create.js [hunk 24]
```diff
@@ -395,14 +392,64 @@ function createParent(params) {
...
+ * @param {Object} [options={}]
+ * @param {string[]} [options.classList=[]] additionnal classes to add to the target
+ * @param {boolean} [options.debug=false] if true, the widget will be appended in
+ *      the DOM. Also, RPCs and uncaught OdooEvent will be logged
+ * @param {Object} [options.events={}] events intercepted by the target.
+ *      key: (event name), value: (callback)
+ * @returns {HTMLElement}
+ */
+function prepareTarget(options={}) {
+    const { classList, debug, events } = options;
```
Comment (aab-odoo) [inline addons/web/static/tests/helpers/test_utils_create.js]: maybe call cleanTarget here

File: addons/web/static/tests/helpers/test_utils_create.js [hunk 25]
```diff
@@ -106,17 +107,15 @@ var createActionManager = async function (params) {
  *   Note that this is particularly useful if you want to intercept events going
  *   up in the init process of the view, because there are no other way to do it
  *   after this method returns
+ *  @param {Boolean} params.doNotDisableAHref will not preventDefault on the A elements of the view if true.
```
Comment (aab-odoo) [inline addons/web/static/tests/helpers/test_utils_create.js]: [params.doNot...=false]

File: addons/web_editor/static/tests/field_html_tests.js [hunk 26]
```diff
@@ -380,7 +380,7 @@ QUnit.module('web_editor', {}, function () {
             form.destroy();
         });
 
-        QUnit.test('save immediately before iframe is rendered in edit mode', async function (assert) {
+        QUnit.skip('save immediately before iframe is rendered in edit mode !!!! THIS TEST DOES NOT WORK !!!!', async function (assert) {
```
Comment (aab-odoo) [inline addons/web_editor/static/tests/field_html_tests.js]: check and remove

File: addons/web/static/tests/main_tests.js [hunk 27]
```diff
@@ -15,6 +16,7 @@ async function startWebClient() {
 
     await session.is_bound;
     session.owlTemplates = session.owlTemplates.replace(/t-transition/g, 'transition');
+    owl.Component.env = makeTestEnvironment();
```
Comment (aab-odoo) [inline addons/web/static/tests/main_tests.js]: move this in first commit

File: addons/web/static/tests/helpers/test_utils_mock.js [hunk 28]
```diff
@@ -11,6 +11,7 @@ odoo.define('web.test_utils_mock', function (require) {
  */
 
 var basic_fields = require('web.basic_fields');
+var concurrency = require('web.concurrency');
```
Comment (aab-odoo) [inline addons/web/static/tests/helpers/test_utils_mock.js]: chelou

File: addons/web/static/tests/main_tests.js (hunk 27 above)
Comment (aab-odoo) [reply addons/web/static/tests/main_tests.js]: but do this in beforeAll

File: addons/web/static/tests/helpers/test_env.js (hunk 8 above)
Comment (Arcasias) [reply addons/web/static/tests/helpers/test_env.js]: Ok

File: addons/web/static/tests/helpers/test_env.js (hunk 11 above)
Comment (Arcasias) [reply addons/web/static/tests/helpers/test_env.js]: Ok

File: addons/web/static/tests/helpers/test_env.js (hunk 9 above)
Comment (Arcasias) [reply addons/web/static/tests/helpers/test_env.js]: Ok

File: addons/web/static/tests/helpers/test_env.js (hunk 12 above)
Comment (Polymorphe57) [reply addons/web/static/tests/helpers/test_env.js]: ok

File: addons/web/static/tests/helpers/test_env.js (hunk 13 above)
Comment (Polymorphe57) [reply addons/web/static/tests/helpers/test_env.js]: ok

File: addons/web_tour/static/src/js/running_tour_action_helper.js (hunk 15 above)
Comment (Polymorphe57) [reply addons/web_tour/static/src/js/running_tour_action_helper.js]: ok

File: addons/web_editor/static/tests/field_html_tests.js (hunk 26 above)
Comment (Polymorphe57) [reply addons/web_editor/static/tests/field_html_tests.js]: ok

File: addons/web/static/tests/helpers/test_utils_create.js (hunk 24 above)
Comment (Polymorphe57) [reply addons/web/static/tests/helpers/test_utils_create.js]: ok

File: addons/web/static/tests/helpers/test_utils.js [hunk 29]
```diff
@@ -122,7 +122,9 @@ return Promise.all([
             intercept: testUtilsMock.intercept,
             patch: testUtilsMock.patch,
             patchDate: testUtilsMock.patchDate,
+            patchRequestAnimationFrame: testUtilsMock.patchRequestAnimationFrame,
```
[the remark this answers is not in the document]
Comment (VincentSchippefilt) [reply addons/web/static/tests/helpers/test_utils.js]: Rien A Foutre ?

[the remark this answers is not in the document]
Comment (VincentSchippefilt) [reply addons/web/static/tests/helpers/test_utils.js]: Roule à fond ?

[the remark this answers is not in the document]
Comment (Arcasias) [reply addons/web/static/tests/helpers/test_utils.js]: Raviollis au fromage

**Comment by VincentSchippefilt** [comment]
merged with https://github.com/odoo/odoo/pull/39491
